server/src/main/java/com/alibaba/fescar/server/coordinator/DefaultCore.java

```diff
@@ -284,18 +289,13 @@ public void doGlobalRollback(GlobalSession globalSession, boolean retrying) thro
                 LOGGER.info("Exception rollbacking branch " + branchSession, ex);
                 if (!retrying) {
                     queueToRetryRollback(globalSession);
-                    if (ex instanceof TransactionException) {
-                        throw (TransactionException)ex;
-                    } else {
-                        throw new TransactionException(ex);
-                    }
                 }
-
+                throw new TransactionException(ex);
             }
 
         }
         if (globalSession.hasBranch()) {
-            SessionHelper.endRollbackFailed(globalSession);
+            throw new ShouldNeverHappenException("GlobalRollback error,GID:" + globalSession.getTransactionId());
```

Comment:
why docommit code is 
`if (globalSession.hasBranch()) {
            LOGGER.info("Global[{}] committing is NOT done.", globalSession.getTransactionId());
            return;
        }`
but doRollback code is {
 if (globalSession.hasBranch()) {
            SessionHelper.endRollbackFailed(globalSession);
}
One has only Info log, and the other has a failed state. Should the commit also fail？

Reply:
now doRollback is throw exx; doGlobalCommit contains asyn commit, so it can have branch.

Reply:
if branchSesson status is  BranchStatus.PhaseOne_Failed, It will  throw an exception